### ISSUE TITLE
Incorrect generated code for generic struct with union

### ISSUE THREAD

**Issue by as-com** [capnproto/capnproto-rust#250]
I tried the following code:

src/test.capnp:
```capnp
@0xef09ef749da18e77;

struct Foo(T) {
    foo @0 :T;
    bar :union {
        baz @1 :Void;
        qux @2 :Int32;
    }
}
```

src/main.rs:
```rust
mod test_capnp;

fn main() {
    println!("Hello, world!");
}
```

cargo.toml:
```toml
[package]
name = "capnpc-test"
version = "0.1.0"
edition = "2021"

[dependencies]
capnp = "0.14"

[build-dependencies]
capnpc = "0.14"

```

build.rs:
```rust
fn main() {
    capnpc::CompilerCommand::new()
        .src_prefix("./src")
        .file("./src/test.capnp")
        .output_path("./src")
        .run()
        .expect("schema compiler command");
}
```

Expected result: the program compiles.
Actual result: 
```
   Compiling capnp v0.14.4
   Compiling capnpc-test v0.1.0 (/home/andrew/dev/capnpc-test)
error[E0107]: this struct takes 1 generic argument but 0 generic arguments were supplied
   --> src/test_capnp.rs:65:58
    |
65  |     pub fn get_bar(self) -> crate::test_capnp::foo::bar::Reader<'a> {
    |                                                          ^^^^^^ expected 1 generic argument
    |
note: struct defined here, with 1 generic parameter: `T`
   --> src/test_capnp.rs:187:16
    |
187 |     pub struct Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    |                ^^^^^^    -
help: add missing generic argument
    |
65  |     pub fn get_bar(self) -> crate::test_capnp::foo::bar::Reader<'a, T> {
    |                                                                   +++

error[E0107]: this struct takes 1 generic argument but 0 generic arguments were supplied
   --> src/test_capnp.rs:141:58
    |
141 |     pub fn get_bar(self) -> crate::test_capnp::foo::bar::Builder<'a> {
    |                                                          ^^^^^^^ expected 1 generic argument
    |
note: struct defined here, with 1 generic parameter: `T`
   --> src/test_capnp.rs:246:16
    |
246 |     pub struct Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    |                ^^^^^^^    -
help: add missing generic argument
    |
141 |     pub fn get_bar(self) -> crate::test_capnp::foo::bar::Builder<'a, T> {
    |                                                                    +++

error[E0107]: this struct takes 1 generic argument but 0 generic arguments were supplied
   --> src/test_capnp.rs:145:61
    |
145 |     pub fn init_bar(self, ) -> crate::test_capnp::foo::bar::Builder<'a> {
    |                                                             ^^^^^^^ expected 1 generic argument
    |
note: struct defined here, with 1 generic parameter: `T`
   --> src/test_capnp.rs:246:16
    |
246 |     pub struct Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    |                ^^^^^^^    -
help: add missing generic argument
    |
145 |     pub fn init_bar(self, ) -> crate::test_capnp::foo::bar::Builder<'a, T> {
    |                                                                       +++

error[E0107]: missing generics for struct `bar::Pipeline`
   --> src/test_capnp.rs:165:59
    |
165 |     pub fn get_bar(&self) -> crate::test_capnp::foo::bar::Pipeline {
    |                                                           ^^^^^^^^ expected 1 generic argument
    |
note: struct defined here, with 1 generic parameter: `T`
   --> src/test_capnp.rs:324:16
    |
324 |     pub struct Pipeline<T> {
    |                ^^^^^^^^ -
help: add missing generic argument
    |
165 |     pub fn get_bar(&self) -> crate::test_capnp::foo::bar::Pipeline<T> {
    |                                                           ~~~~~~~~~~~

For more information about this error, try `rustc --explain E0107`.
error: could not compile `capnpc-test` due to 4 previous errors
```

<details>
<summary>Generated file `src/test_capnp.rs`:</summary>

```rust
// @generated by the capnpc-rust plugin to the Cap'n Proto schema compiler.
// DO NOT EDIT.
// source: test.capnp


pub mod foo { /* T */
  #[derive(Copy, Clone)]
  pub struct Owned<T> {
    _phantom: ::core::marker::PhantomData<T>
  }
  impl <'a, T> ::capnp::traits::Owned<'a> for Owned <T> where T: for<'c> ::capnp::traits::Owned<'c>  { type Reader = Reader<'a, T>; type Builder = Builder<'a, T>; }
  impl <'a, T> ::capnp::traits::OwnedStruct<'a> for Owned <T> where T: for<'c> ::capnp::traits::Owned<'c>  { type Reader = Reader<'a, T>; type Builder = Builder<'a, T>; }
  impl <T> ::capnp::traits::Pipelined for Owned<T> where T: for<'c> ::capnp::traits::Owned<'c>  { type Pipeline = Pipeline<T>; }

  #[derive(Clone, Copy)]
  pub struct Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    reader: ::capnp::private::layout::StructReader<'a>,
    _phantom: ::core::marker::PhantomData<T>
  }

  impl <'a,T> ::capnp::traits::HasTypeId for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    #[inline]
    fn type_id() -> u64 { _private::TYPE_ID }
  }
  impl <'a,T> ::capnp::traits::FromStructReader<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn new(reader: ::capnp::private::layout::StructReader<'a>) -> Reader<'a,T> {
      Reader { reader, _phantom: ::core::marker::PhantomData, }
    }
  }

  impl <'a,T> ::capnp::traits::FromPointerReader<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn get_from_pointer(reader: &::capnp::private::layout::PointerReader<'a>, default: ::core::option::Option<&'a [capnp::Word]>) -> ::capnp::Result<Reader<'a,T>> {
      ::core::result::Result::Ok(::capnp::traits::FromStructReader::new(reader.get_struct(default)?))
    }
  }

  impl <'a,T> ::capnp::traits::IntoInternalStructReader<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn into_internal_struct_reader(self) -> ::capnp::private::layout::StructReader<'a> {
      self.reader
    }
  }

  impl <'a,T> ::capnp::traits::Imbue<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn imbue(&mut self, cap_table: &'a ::capnp::private::layout::CapTable) {
      self.reader.imbue(::capnp::private::layout::CapTableReader::Plain(cap_table))
    }
  }

  impl <'a,T> Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    pub fn reborrow(&self) -> Reader<'_,T> {
      Reader { .. *self }
    }

    pub fn total_size(&self) -> ::capnp::Result<::capnp::MessageSize> {
      self.reader.total_size()
    }
    #[inline]
    pub fn get_foo(self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Reader> {
      ::capnp::traits::FromPointerReader::get_from_pointer(&self.reader.get_pointer_field(0), ::core::option::Option::None)
    }
    pub fn has_foo(&self) -> bool {
      !self.reader.get_pointer_field(0).is_null()
    }
    #[inline]
    pub fn get_bar(self) -> crate::test_capnp::foo::bar::Reader<'a> {
      ::capnp::traits::FromStructReader::new(self.reader)
    }
  }

  pub struct Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    builder: ::capnp::private::layout::StructBuilder<'a>,
    _phantom: ::core::marker::PhantomData<T>
  }
  impl <'a,T> ::capnp::traits::HasStructSize for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    #[inline]
    fn struct_size() -> ::capnp::private::layout::StructSize { _private::STRUCT_SIZE }
  }
  impl <'a,T> ::capnp::traits::HasTypeId for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    #[inline]
    fn type_id() -> u64 { _private::TYPE_ID }
  }
  impl <'a,T> ::capnp::traits::FromStructBuilder<'a> for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn new(builder: ::capnp::private::layout::StructBuilder<'a>) -> Builder<'a, T> {
      Builder { builder, _phantom: ::core::marker::PhantomData, }
    }
  }

  impl <'a,T> ::capnp::traits::ImbueMut<'a> for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn imbue_mut(&mut self, cap_table: &'a mut ::capnp::private::layout::CapTable) {
      self.builder.imbue(::capnp::private::layout::CapTableBuilder::Plain(cap_table))
    }
  }

  impl <'a,T> ::capnp::traits::FromPointerBuilder<'a> for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn init_pointer(builder: ::capnp::private::layout::PointerBuilder<'a>, _size: u32) -> Builder<'a,T> {
      ::capnp::traits::FromStructBuilder::new(builder.init_struct(_private::STRUCT_SIZE))
    }
    fn get_from_pointer(builder: ::capnp::private::layout::PointerBuilder<'a>, default: ::core::option::Option<&'a [capnp::Word]>) -> ::capnp::Result<Builder<'a,T>> {
      ::core::result::Result::Ok(::capnp::traits::FromStructBuilder::new(builder.get_struct(_private::STRUCT_SIZE, default)?))
    }
  }

  impl <'a,T> ::capnp::traits::SetPointerBuilder for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    fn set_pointer_builder<'b>(pointer: ::capnp::private::layout::PointerBuilder<'b>, value: Reader<'a,T>, canonicalize: bool) -> ::capnp::Result<()> { pointer.set_struct(&value.reader, canonicalize) }
  }

  impl <'a,T> Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
    pub fn into_reader(self) -> Reader<'a,T> {
      ::capnp::traits::FromStructReader::new(self.builder.into_reader())
    }
    pub fn reborrow(&mut self) -> Builder<'_,T> {
      Builder { .. *self }
    }
    pub fn reborrow_as_reader(&self) -> Reader<'_,T> {
      ::capnp::traits::FromStructReader::new(self.builder.into_reader())
    }

    pub fn total_size(&self) -> ::capnp::Result<::capnp::MessageSize> {
      self.builder.into_reader().total_size()
    }
    #[inline]
    pub fn get_foo(self) -> ::capnp::Result<<T as ::capnp::traits::Owned<'a>>::Builder> {
      ::capnp::traits::FromPointerBuilder::get_from_pointer(self.builder.get_pointer_field(0), ::core::option::Option::None)
    }
    #[inline]
    pub fn initn_foo(self, length: u32) -> <T as ::capnp::traits::Owned<'a>>::Builder {
      ::capnp::any_pointer::Builder::new(self.builder.get_pointer_field(0)).initn_as(length)
    }
    #[inline]
    pub fn set_foo(&mut self, value: <T as ::capnp::traits::Owned<'_>>::Reader) -> ::capnp::Result<()> {
      ::capnp::traits::SetPointerBuilder::set_pointer_builder(self.builder.get_pointer_field(0), value, false)
    }
    #[inline]
    pub fn init_foo(self, ) -> <T as ::capnp::traits::Owned<'a>>::Builder {
      ::capnp::any_pointer::Builder::new(self.builder.get_pointer_field(0)).init_as()
    }
    pub fn has_foo(&self) -> bool {
      !self.builder.get_pointer_field(0).is_null()
    }
    #[inline]
    pub fn get_bar(self) -> crate::test_capnp::foo::bar::Builder<'a> {
      ::capnp::traits::FromStructBuilder::new(self.builder)
    }
    #[inline]
    pub fn init_bar(self, ) -> crate::test_capnp::foo::bar::Builder<'a> {
      self.builder.set_data_field::<u16>(0, 0);
      self.builder.set_data_field::<i32>(1, 0i32);
      ::capnp::traits::FromStructBuilder::new(self.builder)
    }
  }

  pub struct Pipeline<T> {
    _typeless: ::capnp::any_pointer::Pipeline,
    _phantom: ::core::marker::PhantomData<T>
  }
  impl<T> ::capnp::capability::FromTypelessPipeline for Pipeline<T> {
    fn new(typeless: ::capnp::any_pointer::Pipeline) -> Pipeline<T> {
      Pipeline { _typeless: typeless, _phantom: ::core::marker::PhantomData, }
    }
  }
  impl<T> Pipeline<T> where T: ::capnp::traits::Pipelined, <T as ::capnp::traits::Pipelined>::Pipeline: ::capnp::capability::FromTypelessPipeline  {
    pub fn get_foo(&self) -> <T as ::capnp::traits::Pipelined>::Pipeline {
      ::capnp::capability::FromTypelessPipeline::new(self._typeless.get_pointer_field(0))
    }
    pub fn get_bar(&self) -> crate::test_capnp::foo::bar::Pipeline {
      ::capnp::capability::FromTypelessPipeline::new(self._typeless.noop())
    }
  }
  mod _private {
    use capnp::private::layout;
    pub const STRUCT_SIZE: layout::StructSize = layout::StructSize { data: 1, pointers: 1 };
    pub const TYPE_ID: u64 = 0x9220_f7c6_3b3a_07f3;
  }

  pub mod bar { /* T */
    pub use self::Which::{Baz,Qux};

    #[derive(Copy, Clone)]
    pub struct Owned<T> {
      _phantom: ::core::marker::PhantomData<T>
    }
    impl <'a, T> ::capnp::traits::Owned<'a> for Owned <T> where T: for<'c> ::capnp::traits::Owned<'c>  { type Reader = Reader<'a, T>; type Builder = Builder<'a, T>; }
    impl <'a, T> ::capnp::traits::OwnedStruct<'a> for Owned <T> where T: for<'c> ::capnp::traits::Owned<'c>  { type Reader = Reader<'a, T>; type Builder = Builder<'a, T>; }
    impl <T> ::capnp::traits::Pipelined for Owned<T> where T: for<'c> ::capnp::traits::Owned<'c>  { type Pipeline = Pipeline<T>; }

    #[derive(Clone, Copy)]
    pub struct Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      reader: ::capnp::private::layout::StructReader<'a>,
      _phantom: ::core::marker::PhantomData<T>
    }

    impl <'a,T> ::capnp::traits::HasTypeId for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      #[inline]
      fn type_id() -> u64 { _private::TYPE_ID }
    }
    impl <'a,T> ::capnp::traits::FromStructReader<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn new(reader: ::capnp::private::layout::StructReader<'a>) -> Reader<'a,T> {
        Reader { reader, _phantom: ::core::marker::PhantomData, }
      }
    }

    impl <'a,T> ::capnp::traits::FromPointerReader<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn get_from_pointer(reader: &::capnp::private::layout::PointerReader<'a>, default: ::core::option::Option<&'a [capnp::Word]>) -> ::capnp::Result<Reader<'a,T>> {
        ::core::result::Result::Ok(::capnp::traits::FromStructReader::new(reader.get_struct(default)?))
      }
    }

    impl <'a,T> ::capnp::traits::IntoInternalStructReader<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn into_internal_struct_reader(self) -> ::capnp::private::layout::StructReader<'a> {
        self.reader
      }
    }

    impl <'a,T> ::capnp::traits::Imbue<'a> for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn imbue(&mut self, cap_table: &'a ::capnp::private::layout::CapTable) {
        self.reader.imbue(::capnp::private::layout::CapTableReader::Plain(cap_table))
      }
    }

    impl <'a,T> Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      pub fn reborrow(&self) -> Reader<'_,T> {
        Reader { .. *self }
      }

      pub fn total_size(&self) -> ::capnp::Result<::capnp::MessageSize> {
        self.reader.total_size()
      }
      #[inline]
      pub fn which(self) -> ::core::result::Result<WhichReader, ::capnp::NotInSchema> {
        match self.reader.get_data_field::<u16>(0) {
          0 => {
            ::core::result::Result::Ok(Baz(
              ()
            ))
          }
          1 => {
            ::core::result::Result::Ok(Qux(
              self.reader.get_data_field::<i32>(1)
            ))
          }
          x => ::core::result::Result::Err(::capnp::NotInSchema(x))
        }
      }
    }

    pub struct Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      builder: ::capnp::private::layout::StructBuilder<'a>,
      _phantom: ::core::marker::PhantomData<T>
    }
    impl <'a,T> ::capnp::traits::HasStructSize for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      #[inline]
      fn struct_size() -> ::capnp::private::layout::StructSize { _private::STRUCT_SIZE }
    }
    impl <'a,T> ::capnp::traits::HasTypeId for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      #[inline]
      fn type_id() -> u64 { _private::TYPE_ID }
    }
    impl <'a,T> ::capnp::traits::FromStructBuilder<'a> for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn new(builder: ::capnp::private::layout::StructBuilder<'a>) -> Builder<'a, T> {
        Builder { builder, _phantom: ::core::marker::PhantomData, }
      }
    }

    impl <'a,T> ::capnp::traits::ImbueMut<'a> for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn imbue_mut(&mut self, cap_table: &'a mut ::capnp::private::layout::CapTable) {
        self.builder.imbue(::capnp::private::layout::CapTableBuilder::Plain(cap_table))
      }
    }

    impl <'a,T> ::capnp::traits::FromPointerBuilder<'a> for Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn init_pointer(builder: ::capnp::private::layout::PointerBuilder<'a>, _size: u32) -> Builder<'a,T> {
        ::capnp::traits::FromStructBuilder::new(builder.init_struct(_private::STRUCT_SIZE))
      }
      fn get_from_pointer(builder: ::capnp::private::layout::PointerBuilder<'a>, default: ::core::option::Option<&'a [capnp::Word]>) -> ::capnp::Result<Builder<'a,T>> {
        ::core::result::Result::Ok(::capnp::traits::FromStructBuilder::new(builder.get_struct(_private::STRUCT_SIZE, default)?))
      }
    }

    impl <'a,T> ::capnp::traits::SetPointerBuilder for Reader<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      fn set_pointer_builder<'b>(pointer: ::capnp::private::layout::PointerBuilder<'b>, value: Reader<'a,T>, canonicalize: bool) -> ::capnp::Result<()> { pointer.set_struct(&value.reader, canonicalize) }
    }

    impl <'a,T> Builder<'a,T> where T: for<'c> ::capnp::traits::Owned<'c>  {
      pub fn into_reader(self) -> Reader<'a,T> {
        ::capnp::traits::FromStructReader::new(self.builder.into_reader())
      }
      pub fn reborrow(&mut self) -> Builder<'_,T> {
        Builder { .. *self }
      }
      pub fn reborrow_as_reader(&self) -> Reader<'_,T> {
        ::capnp::traits::FromStructReader::new(self.builder.into_reader())
      }

      pub fn total_size(&self) -> ::capnp::Result<::capnp::MessageSize> {
        self.builder.into_reader().total_size()
      }
      #[inline]
      pub fn set_baz(&mut self, _value: ())  {
        self.builder.set_data_field::<u16>(0, 0);
      }
      #[inline]
      pub fn set_qux(&mut self, value: i32)  {
        self.builder.set_data_field::<u16>(0, 1);
        self.builder.set_data_field::<i32>(1, value);
      }
      #[inline]
      pub fn which(self) -> ::core::result::Result<WhichBuilder, ::capnp::NotInSchema> {
        match self.builder.get_data_field::<u16>(0) {
          0 => {
            ::core::result::Result::Ok(Baz(
              ()
            ))
          }
          1 => {
            ::core::result::Result::Ok(Qux(
              self.builder.get_data_field::<i32>(1)
            ))
          }
          x => ::core::result::Result::Err(::capnp::NotInSchema(x))
        }
      }
    }

    pub struct Pipeline<T> {
      _typeless: ::capnp::any_pointer::Pipeline,
      _phantom: ::core::marker::PhantomData<T>
    }
    impl<T> ::capnp::capability::FromTypelessPipeline for Pipeline<T> {
      fn new(typeless: ::capnp::any_pointer::Pipeline) -> Pipeline<T> {
        Pipeline { _typeless: typeless, _phantom: ::core::marker::PhantomData, }
      }
    }
    impl<T> Pipeline<T> where T: ::capnp::traits::Pipelined, <T as ::capnp::traits::Pipelined>::Pipeline: ::capnp::capability::FromTypelessPipeline  {
    }
    mod _private {
      use capnp::private::layout;
      pub const STRUCT_SIZE: layout::StructSize = layout::StructSize { data: 1, pointers: 1 };
      pub const TYPE_ID: u64 = 0xc558_9057_3aad_b32f;
    }
    pub enum Which {
      Baz(()),
      Qux(i32),
    }
    pub type WhichReader = Which;
    pub type WhichBuilder = Which;
  }
}
```

</details>

It appears there is incorrect handling of this case, where capnpc is missing the generic parameter for the union getters/setters.

**Comment by comath** [comment]
I think it's incorrectly detecting that the union is generic, the parameter shouldn't be there in the generated code for `bar`. So, it's not a missing parameter, it's extras. The `T` is showing up where it shouldn't.

**Comment by comath** [comment]
Though, If you edit your `test.capnp` to make `bar` generic:

The intermediary schema file generated for `bar` is as expected (you get an extra accessor), but there's no difference in the schema for Foo. 

The accessor for `bar` in `Foo` is still:
```
( name = "bar",
            codeOrder = 1,
            discriminantValue = 65535,
            group = (
              typeId = 14220274527743357743 ),
            ordinal = (implicit = void) ) ] ),
```
So, no hint that the nested union is generic.

**Comment by dwrensha** [comment]
A somewhat simpler case that also fails:
```capnp
struct Foo(T) {
    bar :group {
        baz @0 :T;
    }
}
```

Looks like we need to add generic parameters here: https://github.com/capnproto/capnproto-rust/blob/267b8b8939fb63b1765741617b98d1029466c917/capnpc/src/codegen.rs#L514-L517

**Comment by dwrensha** [comment]
Potential fix here: #252. (Needs some cleanup before landing.)

**Comment by dwrensha** [comment]
Thanks for the report!
